sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs

```diff
@@ -145,57 +111,49 @@ internal void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockSt
-                    }
-                }, symbolKinds);
+            RegisterContextAction(act => context.RegisterSymbolAction(act, symbolKinds), action, c => c.GetSyntaxTree());
 
-        internal static bool IsAnalysisDisabled(SyntaxTree tree, IEnumerable<DiagnosticDescriptor> supportedDiagnostics = null)
+        private void RegisterContextAction<TContext>(Action<Action<TContext>> registrationAction, Action<TContext> registeredAction,
+            Func<TContext, SyntaxTree> getSyntaxTree)
         {
-            // This is the new way SonarLint will handle whether or not the analysis should be performed...
-            // (checking `supportedDiagnostics != null` is to force providing the old behavior for SonarCodeFixProvider)
-            if (ShouldExecuteRuleFunc != null &&
-                supportedDiagnostics != null)
+            /*
+             * For performance reasons, we don't want to register callbacks for rules that should not be run. However, we don't
+             * necessarily have enough information to make the decision up front, so the check is broken into two parts:
+             *
+             * 1.   Pre-registration: at this point we only know the rules that are being registered. If we can decide from the
+             *      diagnostic descriptors that a rule should not be run then we won't register the callback with the analysis
+             *      context.
+             *
+             * 2.   Post-registration: we are now being called back by the analysis context, so we now have access to the syntax
+             *      tree (and project) that are being analyzed, and can use that information to decide whether to execute the
+             *      logic of the rule.
+             */
+
+            if (ShouldRegisterContextAction?.Invoke(this.supportedDiagnostics) != false)
             {
-                Debug.Assert(ShouldAnalysisBeDisabled == null, "Not expecting SonarLint to set both the old and the new" +
-                    "delegates.");
-
-                // If any of the diagnostic is enabled then allow to run the rule action BUT filter out at the time when the
-                // issue is being reported.
-                return !ShouldExecuteRuleFunc(new AnalysisRunContext(tree, supportedDiagnostics));
+                registrationAction(
+                    c =>
+                    {
+                        if (IsRegisteredActionEnabled(getSyntaxTree(c)))
+                        {
+                            registeredAction(c);
+                        }
+                    });
             }
-
-            // ... but for compatibility purposes we need to keep handling the old-fashioned way
-            return tree != null &&
-                ShouldAnalysisBeDisabled != null &&
-                ShouldAnalysisBeDisabled(tree);
         }
+
+        internal static bool IsRegisteredActionEnabled(SyntaxTree tree) =>
+            ShouldExecuteRegisteredAction == null ||
+            tree == null ||

```

Comment:
I am not sure why we want to execute the action when the syntax tree is null. I wouldn't have done that but it is there and I am not sure of the impact.

Reply:
If the tree is ever null then it means there's a bug somewhere, either in our code or in Roslyn. I guess our main concern at that point is just coding defensively so we don't throw.

Reply:
But with this code if the tree is null then we DO call the action while for me we shouldn't.

Reply:
As long as we don't throw I think it's ok to do either.